### PR TITLE
Refactor metadata tool to use deps.dev and libraries.io

### LINKED ISSUE #335: Implement metadata via osv.dev and deps.dev endpoints
We never really implemented metadata gatherers. For supported types, we should consider just querying [osv.dev](https://osv.dev/#use-the-api) (for vulnerability information) and [deps.dev](https://deps.dev) (for metadata). We could also (if the user supplies an API key) query [libraries.io](https://libraries.io/api).

Example for osv.dev:
```
curl -X POST -d   '{"version": "2.4.1", "package": {"name": "jinja2", "ecosystem": "PyPI"}}' https://api.osv.dev/v1/query" | jq
{
  "vulns": [
    {
      "id": "PYSEC-2014-8",
      "details": "The default configuration for bccache.FileSystemBytecodeCache in Jinja2 before 2.7.2 does not properly create temporary files, which allows local users to gain privileges via a crafted .cache file with a name starting with __jinja2_ in /tmp.",
      "aliases": [
        "CVE-2014-1402"
      ],
      "modified": "2021-07-05T00:01:22.043149Z",
      "published": "2014-05-19T14:55:00Z",
      "references": [
        {
          "type": "WEB",
          "url": "https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=734747"
        },
...
```

Example for deps.dev:
```
https://deps.dev/_/s/npm/p/left-pad/v/1.3.0
{
  "package": {
    "system": "NPM",
    "name": "left-pad"
  },
  "owners": [
    "cameron.westland@autodesk.com",
    "maochenyan@gmail.com",
    "sebmck@gmail.com"
  ],
  "version": {
    "version": "1.3.0",
    "symbolicVersions": [
      "latest"
    ],
    "createdAt": 1523236245,
    "refreshedAt": 1635443183,
    "isDefault": true,
    "deprecated": "use String.prototype.padStart()",
    "description": "String left pad",
...
```

### PR DESCRIPTION
**Pull request by scovetta**
Fixes #335 

This PR adds support for querying the deps.dev API (unauthenticated) and libraries.io (unauthenticated or authenticated via a user-supplied API token). The results are serialized (JSON) and written to the console.

### CODE REVIEW

**Comment by gfs** [comment]
Looks like the test might need a back off or to make fewer requests  (use mocks?) - there's at least one test that seems to be dialing with rate limit.

File: src/oss-metadata/MetadataTool.cs [hunk 1]
```diff
@@ -22,52 +26,40 @@ public static IEnumerable<Example> Examples
...
-                    metadata = await projectManager.GetPackageMetadataAsync(purl, useCache);
-                }
-            }
-            catch (Exception ex)
-            {
-                Logger.Warn(ex, "Error identifying source repository for {0}: {1}", purl, ex.Message);
-            }
-
-            return metadata;
-        }
```
[the remark this answers is not in the document]
Comment (scovetta) [reply src/oss-metadata/MetadataTool.cs:69]: We already have the concept of metadata sources (deps.dev is default, libraries.io can be specified) -- let's add "native" and delegate those out to the package managers that we have native implementations for.

And yes, for each one, we can staple in a "source" field that specifies metadata about the metadata.

**Comment by gfs** [comment]
That sounds good to me.

**Comment by scovetta** [comment]
I added a NativeMetadataSource class which just delegates out to GetMetadataAsync where implemented.

I went with JsonDocument as the return type -- it allows us to specify a JMESPath expression on the command line and do some filtering/carving. But native sources don't necessarily emit JSON -- in the case of Hackage, it looks like HTML (unless it's broken). So if the tool can't parse JSON, it just sticks the string inside a JSON envelope. It's not great -- maybe we should have some sort of intermediate object that we use instead of JsonDocument.

**Comment by gfs** [comment]
I think ideally we would define a native meta data object and just only populate the fields that that particular native metadata resolver can populate. I think the overall objective is to deal with any type of package in a non-repo specific way so having a package type specific jsondoc isn't super user friendly.  That doesn't have to be a blocker for this PR, but the goal should be a Metadata object with nulls for the fields that native resolver doesn't support imo.

**Comment by scovetta** [comment]
Once this PR is ready, it should fix #288.

**Comment by scovetta** [comment]
Will anyone have heartburn over merging this PR (once the tests pass) and improve the metadata in the near future (including, potentially normalizing the data, which will take a good bit of work)?

**Comment by gfs** [comment]
Fine with me to merge (though 2 tests are failing so those should be addressed first) and handle the larger metadata issue at a later date.

**Comment by gfs** [comment]
@scovetta Do you want me to resolve the test failures and push a new version?

**Comment by scovetta** [comment]
> @scovetta Do you want me to resolve the test failures and push a new version?

Please and thank you!

**Comment by jpinz** [comment]
> Will anyone have heartburn over merging this PR (once the tests pass) and improve the metadata in the near future (including, potentially normalizing the data, which will take a good bit of work)?

We had talked about metadata normalization a lot in the Terrapin team. Let me know if you want to find time soon for us to all chat about what we were thinking about.

**Comment by gfs** [comment]
@jpinz that sounds like a good idea. Please go ahead and set that meeting up. 

@scovetta no problem. ill work to get this resolved next week.

**Comment by gfs** [comment]
Rerun locally and tests pass. Might just be an intermittent download test issue - but the records are gone because the run was long enough ago. Attempting to rerun pipeline

/azp run

**Comment by gfs** [comment]
Checks now passing. Looks like just an intermittent test issue.

**Review by gfs — APPROVED** [review]

**Pull request merged**